cpp/src/groupby/common/kernel_utils.hpp

```diff
@@ -0,0 +1,224 @@
+ *
+ * For an invalid combination of SourceType and operator,
+ *`target_type_t<SourceType, operator>` yields a `void` TargetType. This
+ * specialization will be invoked for any invalid combination and cause a
+ * runtime error.
+ *
+ * @note A struct is used instead of a function to allow for partial
+ * specialization.
+ *---------------------------------------------------------------------------**/
+template <typename SourceType, operators op, bool values_have_nulls,
+          typename Enable = void>
+struct update_target_element {
+  __device__ inline void operator()(gdf_column const& target,
+                                    gdf_size_type target_index,
+                                    gdf_column const& source,
+                                    gdf_size_type source_index) {
+    release_assert(false && "Invalid Source type and Aggregation combination.");
+  }
+};
+
+/**---------------------------------------------------------------------------*
+ * @brief Specialization for valid SourceType and op combinations.
+ *
+ * @tparam SourceType Type of the source element
+ * @tparam op The operation to perform
+ * @tparama values_have_nulls Indicates the potential for null values in the
+ * source
+ *---------------------------------------------------------------------------**/
+template <typename SourceType, operators op, bool source_has_nulls>
+struct update_target_element<
+    SourceType, op, source_has_nulls,
+    std::enable_if_t<not std::is_void<target_type_t<SourceType, op>>::value>> {
+  /**---------------------------------------------------------------------------*
+   * @brief Performs in-place update of a target element via a binary operation
+   * with a source element.
+   *
+   * @note It is assumed the source element is not NULL, i.e., a NULL source
+   * element should be detected before calling this function.
+   *
+   * @note If `source_has_nulls==true`, it is assumed that `target` is nullable
+   *
+   * If the target element is NULL, it is assumed that the target element was
+   * initialized with the identity of the aggregation operation. The target is
+   * updated with the result of the aggregation with the source element, and the
+   * target column's bitmask is updated to indicate the target element is no
+   * longer NULL.
+   *
+   * @param target Column containing target element
+   * @param target_index Index of target element
+   * @param source Column containing source element
+   * @param source_index Index of source element
+   *---------------------------------------------------------------------------**/
+  __device__ inline void operator()(gdf_column const& target,
+                                    gdf_size_type target_index,
+                                    gdf_column const& source,
+                                    gdf_size_type source_index) {
+    using TargetType = target_type_t<SourceType, op>;
+    
+    assert(gdf_dtype_of<TargetType>() == target.dtype); 
```

Comment:
```suggestion
    CUDF_EXPECTS(gdf_dtype_of<TargetType>() == target.dtype); 
```

Reply:
This is device code. You can't use `CUDF_EXPECTS`.

Reply:
Why not expand `CUDF_EXPECTS` to include `release_assert` and an ifdef switch to choose between exception and `release_assert` based on whether it's device code or host.

Reply:
Hmmm. I think it would be better to CUDF_EXPECTS in the function that launches the kernel, rather than do either assert or EXPECTS inside the kernel.

Reply:
> Why not expand CUDF_EXPECTS to include release_assert and an ifdef switch to choose between exception and release_assert based on whether it's device code or host.

Yes, I've been working on that, but there's a few snags that I'm ironing out. That said, it would be far too expensive to do a `release_assert` here. The point of an `assert` is for extra safety checking in debug mode. 

> I think it would be better to CUDF_EXPECTS in the function that launches the kernel, rather than do either assert or EXPECTS inside the kernel.

Yes, that is the expectation. Though I don't have control over who calls the operator. The `assert` is just a sanity check in case someone is debugging.

Reply:
> it would be far too expensive to do a `release_assert` here. The point of an `assert` is for extra safety checking in debug mode.

I'm only saying this because I saw this in `type_dispatcher` and have used it as reference for what to do when you want to `EXPECT` in `__host__ __device__` function.

Reply:
So. No changes here. I will mark this as a resolved.